torch/nested/_internal/sdpa.py

```diff
@@ -736,9 +736,30 @@ def jagged_scaled_dot_product_attention(
             attention.squeeze(0), output_nt_info["offsets"]
         ).transpose(1, 2)
     elif backend_choice == SDPBackend.MATH:
-        return torch._scaled_dot_product_attention_math(
+        # convert jagged layout Nested Tensor to strided layout Nested Tensor
+        # which support the math implementation of SDPA
+        def get_strided_layout_nested_tensor(jagged_layout_nt):
+            lengths = jagged_layout_nt._offsets[1:] - jagged_layout_nt._offsets[:-1]
+            transpose = torch.transpose(jagged_layout_nt, 1, 2)
+            tensor_list = buffer_from_jagged(transpose).split(list(lengths), dim=0)
+            c_nt = torch.nested.as_nested_tensor(list(tensor_list))
+            c_nt = c_nt.transpose(1, 2).contiguous()
+            return c_nt
+
+        query = get_strided_layout_nested_tensor(query)
+        key = get_strided_layout_nested_tensor(key)
+        value = get_strided_layout_nested_tensor(value)
+
+        attn_out = torch._scaled_dot_product_attention_math(
             query, key, value, attn_mask, dropout_p, is_causal, scale=scale
         )[0]
+        attn_out = attn_out.transpose(1, 2)
+        attn_out = torch.nested.as_nested_tensor(
+            list(attn_out.unbind()), layout=torch.jagged
+        )
```

Comment:
Could we reshape the output buffer to 2D and then use ViewNestedFromBuffer?
If we construct with as_nested_tensor, it would acquire a new singleton_int and you wouldn't be able to do elementwise binary ops for example because we would consider the two tensors to have different jaggedness

Reply:
I think that's a bit tricky to do so. The c++ NT values doesn't change after transpose. So I can't transpose it and then reshape it to be the buffer for ViewNestedFromBuffer.  

(More context: the output of _scaled_dot_product_attention_math is a 4D C++ Nested tensor with shape like [B, D1, j0, D2].  The goal is to convert this C++ NT to PY NT. Since ViewNestedFromBuffer assume the jagged dimension is the 2nd dimension, I need to transpose the [B, D1, j0, D2] to  [B, j0, D1, D2] first. But since transpose doesn't change the values() in C++ NT, I can't get the transposed values.)

Any other ways to work around it? By the way, for my use case, we patch the attention output before doing any elementwise binary operations, so this is not a blocker for us.

Reply:
What if you got the values from cpp, makes it 2D, and transpose the values yourself and then call contiguous before calling ViewNestedFromBuffer.